Google.PowerShell/Common/ReportCmdletResults.cs

```diff
@@ -0,0 +1,206 @@
+
+        /// <summary>
+        /// Category for all analytic events reported through this instance.
+        /// </summary>
+        private AnalyticsEventCategory _category = AnalyticsEventCategory.CmdletInvocation;
+
+        public GoogleAnalyticsCmdletReporter(string clientID, AnalyticsEventCategory analyticsCategory)
+        {
+            // We report results to Google Analytics if the user has opted-into Cloud SDK metric reporting. However, we
+            // don't want to report data during development. (e.g. when running unit tests.) The data is still sent, but
+            // only to a debug server.
+            string envVar = Environment.GetEnvironmentVariable("DISABLE_POWERSHELL_ANALYTICS");
+            bool disableReporting = !String.IsNullOrWhiteSpace(envVar);
+
+            _category = analyticsCategory;
+            if (s_reporter == null)
+            {
+                // BUG: We only forward the clientID when the AnalyticsReporter is constructed the first time.
+                // If the user opts-out and then opts-into analytics reporting, their client ID will change,
+                // but we will keep using the same cached version. Will persist until the GoogleCloud PowerShell
+                // module is unloaded, i.e. until the PowerShell session terminates.
+                s_reporter = new Lazy<EventsReporter>(() =>
+                {
+                    var analyticsReporter = new AnalyticsReporter(
+                        PropertyId,
+                        clientId: clientID,
+                        appName: ApplicationName,
+                        appVersion: s_appVersion.Value,
+                        debug: disableReporting);
+
+                    return new EventsReporter(analyticsReporter);
+                });
+            }
+        }
+
+        public void ReportSuccess(string cmdletName, string parameterSet)
+        {
+            Report(cmdletName, parameterSet, null);
+        }
+
+        public void ReportFailure(string cmdletName, string parameterSet, int errorCode)
+        {
+            Report(cmdletName, parameterSet, errorCode);
+        }
+
+        private void Report(string cmdletName, string parameterSet, int? errorCode)
+        {
+            // The event type will be CmdletInvocation or ProviderInvocation. The specific cmdlet and parameter set
+            // used are encoded in the event's metadata.
+            AnalyticsEvent eventData = new AnalyticsEvent(
+                _category.ToString(),
+                "cmdletName", cmdletName,
+                "parameterSet", (parameterSet ?? "null"),
+                "errorCode", (errorCode.HasValue ? errorCode.ToString() : "null"));
+
+            s_reporter.Value.ReportEvent(
+                source: "virtual.powershell",
+                eventType: "powershell",
+                eventName: eventData.Name,
+                userLoggedIn: true,
+                projectNumber: null,
```

Comment:
Ack. I'll put a TODO in and wire that in in a future CL. It should be possible for us to get that information from the current Cloud SDK config.

However, some commands cross projects. e.g. you can access GCS bucket foo, which is "owned" by a project different than the one in the current Cloud SDK config. Though, I assume gsutil has the same problem.